File: java/socketd/src/main/java/org/noear/socketd/transport/core/Message.java
package org.noear.socketd.transport.core;


/**
 * 消息
 *
 * @author noear
 * @since 2.0
 */
public interface Message extends Entity{

    /**
     * 是否为请求
     */
    boolean isRequest();

    /**
     * 是否为订阅
     */
    boolean isSubscribe();

    /**
     * 范围开始
     * */
    default int rangeStart(){
        return metaAsInt(EntityMetas.META_RANGE_START);
    }

    /**
     * 范围大小
     * */
    default int rangeSize(){
        return metaAsInt(EntityMetas.META_RANGE_SIZE);
    }

    /**
     * 获取消息流Id（用于消息交互、分片）
     */
    String sid();

    /**
     * 获取消息事件
     */
    String event();

    /**
     * 获取消息实体（有时需要获取实体）
     */
    Entity entity();
}

File: python/socketd/socketd/transport/core/Message.py
from abc import ABC, abstractmethod

from .Entity import Entity, Reply
from .EntityMetas import EntityMetas


class Message(Reply, ABC):
    def at_name(self) -> str:
        return self.meta("@")

    def range_start(self):
        return self.meta_as_int(EntityMetas.META_RANGE_START)

    def range_size(self):
        return self.meta_as_int(EntityMetas.META_RANGE_SIZE)

    @abstractmethod
    def is_request(self) -> bool:
        ...

    @abstractmethod
    def is_subscribe(self) -> bool:
        ...

    @abstractmethod
    def event(self) -> str:
        ...

    @abstractmethod
    def entity(self) -> Entity:
        ...


class MessageInternal(Message, ABC):
    @abstractmethod
    def flag(self) -> int:
        ...
